fero/analysis.py

```diff
@@ -148,3 +328,314 @@ def make_prediction(
+
+    def _verify_cost_goal(self, goal: dict):
+        """Verify that a cost goal is correct"""
+        for factor in goal["cost_function"]:
+            factor_name = factor["name"]
+            factor_type = self._get_factor_dtype(factor_name)
+            if factor_type is None:
+                raise FeroError(f'"{factor_name}" is not a factor')
+            # Implicitly find missing factors
+            if factor_type != "float":
+                raise FeroError("Cost functions factors must be factors")
+
+    def _verify_constraints(self, constraints: List[dict]):
+        """verify provided constraints are in the analysis"""
+        all_names = self.factor_names + self.target_names
+        for constraint in constraints:
+            constraint_name = constraint["name"]
+            if constraint_name not in all_names:
+                raise FeroError(
+                    f'Constraint "{constraint_name}" is not in this analysis'
+                )
+
+    def _cross_verify_optimization(self, goal: dict, constraints: List[dict]):
+        """Verify the config has the correct combined targets, costs and factors"""
+        is_cost = goal.get("type", None) == "COST"
+
+        cost_factors = []
+        goal_factor = []
+        target_factor = []
+        constraint_factors = []
+        constraint_targets = []
+
+        if is_cost:
+            for factor in goal["cost_function"]:
+                cost_factors.append(factor["name"])
+        else:
+            if goal["factor"]["name"] in self.factor_names:
+                goal_factor.append(goal["goal"])
+            else:
+                target_factor.append(goal["goal"])
+
+        for constraint in constraints:
+            if constraint["name"] in self.factor_names:
+                constraint_factors.append(constraint["name"])
+            else:
+                constraint_targets.append(constraint["name"])
+
+        if len(target_factor + constraint_targets) < 1:
+            raise FeroError("No Targets specified")
+
+        if len(constraint_factors + cost_factors + goal_factor) > 3:
+            raise FeroError(
+                "A maximum of three factors can be specified in an optimization"
+            )
+
+    def _verify_fixed_factors(self, fixed_factors: dict):
+        """Check that the provided fixed factors are in the analysis"""
+        all_columns = self.target_names + self.factor_names
+        for key in fixed_factors.keys():
+            if key not in all_columns:
+                raise FeroError(f'"{key}" is not a valid factor')
+
+    def _get_basis_values(self):
+        """Gets median fixed values from presentation data"""
+        reg_data = next(
+            d
+            for d in self._presentation_data["data"]
+            if d["id"] == "regression_simulator"
+        )
+        # yuck massive ugly json
+        base_values = {f["factor"]: f["median"] for f in reg_data["content"]["factors"]}
+
+        target_values = {
```

Comment:
I don't think you actually need to include the target values in the basis list, but I also don't think it hurts.